code/modules/projectiles/ammunition/smartmag.dm

```diff
@@ -0,0 +1,168 @@
+///////// Smart Mags /////////
+
+/obj/item/ammo_magazine/smart
+	name = "smart magazine"
+	icon_state = "smartmag-empty"
+	desc = "A Hephaistos Industries brand Smart Magazine. It uses advanced matter manipulation technology to create bullets from energy. Simply present your loaded gun or magazine to the Smart Magazine."
+	multiple_sprites = 1
+	max_ammo = 20
+	mag_type = MAGAZINE
+
+	caliber = null 	 //Set later
+	ammo_type = null //Set later
+	initial_ammo = 0 //Ensure no problems with no ammo_type or caliber set
+
+	var/production_time = 3 //20 bullets per minute
+	var/last_production_time = 0
+	var/production_cost = null
+	var/production_modifier = 2.5
+
+	var/obj/item/weapon/cell/attached_cell = null
+
+	var/emagged = 0
+
+/obj/item/ammo_magazine/smart/New()
+	processing_objects |= src
+	..()
+
+/obj/item/ammo_magazine/smart/Destroy()
+	processing_objects -= src
+	..()
+
+/obj/item/ammo_magazine/smart/examine(mob/user)
+	..()
+	if(attached_cell)
+		user << "<span class='notice'>\The [src] is loaded with a [attached_cell.name]. It is [round(attached_cell.percent())]% charged.</span>"
+	if(!attached_cell)
+		user << "<span class='warning'>\The [src] does not appear to have a power source installed.</span>"
+
+/obj/item/ammo_magazine/smart/update_icon()
+	if(attached_cell)
+		icon_state = "smartmag-filled"
+	else if(!(attached_cell))
+		icon_state = "smartmag-empty"
+
+/obj/item/ammo_magazine/smart/proc/chargereduction()
+	if(attached_cell)
+		if(attached_cell.checked_use(production_cost))
+			return 1
+		else
+			return 0
+
+	return null
+
+/obj/item/ammo_magazine/smart/proc/set_production_cost(var/obj/item/ammo_casing/A)
+	var/list/matters = ammo_repository.get_materials_from_object(A)
+	var/tempcost
+	for(var/key in matters)
+		var/value = matters[key]
+		tempcost += value * production_modifier
+	production_cost = tempcost
+
+/obj/item/ammo_magazine/smart/attack_self(mob/user)
+	if(emagged)
+		return ..()
+	else
+		user << "<span class='notice'>\The [src] is not designed to be unloaded.</span>"
+		return
+
+/obj/item/ammo_magazine/smart/emag_act(var/remaining_charges, var/mob/user)
+	if(!emagged)
+		user << "<span class='notice'>You overload \the [src]'s security measures causing widespread destabilisation. It is likely you could empty \the [src] now.</span>"
+		emagged = 1
+		return 1
+
+/obj/item/ammo_magazine/smart/emp_act(severity)
+	..()
+	if(attached_cell)
```

Comment:
This actually might end up emping cell twice.
IIRC standard emp_act() is emp self, emp everything in contents - so if attached_cell is in this item, it would emp'd in ..(), and then here.